Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -240,51 +339,68 @@ public struct HTTPHeaders: CustomStringConvertible {
     /// - Parameter name: The header field name whose values are to be retrieved.
     /// - Returns: A list of the values for that header field name.
     public subscript(name: String) -> [String] {
-        if let result = storage[name.lowercased()] {
-            return result.map { tuple in tuple.1 }
+        guard !self.headers.isEmpty else {
+            return []
         }
-        return []
-    }
 
-    /// Serializes this HTTP header block to bytes suitable for writing to the wire.
-    ///
-    /// - Parameter buffer: A buffer to write the serialized bytes into. Will increment
-    ///     the writer index of this buffer.
-    func write(buffer: inout ByteBuffer) {
-        for (key, values) in storage {
-            if key != "set-cookie" {
-                writeListHeaderValues(buffer: &buffer, key: key, values: values)
-            } else {
-                writeSequentialHeaderValues(buffer: &buffer, key: key, values: values)
+        let utf8 = name.utf8
+        let stringLength = utf8.count
+        var array: [String] = []
+        for header in headers {
+            if stringLength == header.name.length && self.buffer.equalCaseInsensitiveASCII(seq: utf8, at: header.name) {
+                array.append(string(idx: header.value))
             }
         }
-        buffer.write(staticString: crlf)
+        return array
     }
 
-    /// Used for most HTTP headers, which can be represented as a single line joined by commas.
-    private func writeListHeaderValues(buffer: inout ByteBuffer, key: String, values: [(String, String)]) {
-        buffer.write(string: key)
-        buffer.write(staticString: headerSeparator)
+    /// Checks if a header is present
+    ///
+    /// - parameters:
+    ///     - name: The name of the header
+    ///     - value: The value of the header or nil if the name should not be checked
+    //  - returns: `true` if a header with the name (and value) exists, `false` otherwise.
+    public func contains(name: String, value: String? = nil) -> Bool {
+        guard !self.headers.isEmpty else {
+            return false
+        }
 
-        var writerIndex = buffer.writerIndex
-        for (_, value) in values {
-            buffer.write(string: value)
-            writerIndex = buffer.writerIndex
-            buffer.write(staticString: ",")
+        let utf8 = name.utf8
+        let stringLength = utf8.count
+        for header in headers {
+            if stringLength == header.name.length && self.buffer.equalCaseInsensitiveASCII(seq: utf8, at: header.name) {
+                if let value = value {
+                    return value.count == header.value.length && self.buffer.getString(at: header.value.start, length: header.value.length) == value
+                }
+                return true
+            }
```

Comment:
There are three methods here that have very similar bodies. We should probably rewrite those as either methods that take non-escaping closures or methods that return the header indices so that we can reduce the code duplication.

Reply:
@Lukasa will review this ... that said while these look kind of similar they are not completely as one is starting from the back and the others not etc.